_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,203 @@
+
+The job graph also depicts various data exchange patterns between the operators.
+
+* The __FORWARD__ connection following the Transaction Source means that all data consumed by one of the parallel instances of the Transaction Source operator is transferred to exactly one instance of the subsequent `DynamicKeyFunction` operator, without redistribution. It also indicates the same level of parallelism of the two connected operators (12 in this case). This communication pattern is depicted in Figure 3. Orange circles represent transactions and dotted rectangles depict parallel instances of the operators (also called [*tasks*](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator)).  
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across tasks: FORWARD"/>
+<br/>
+<i><small>Figure 3: Message passing across tasks: FORWARD</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/forward.png)
+
+* The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using a `keyBy` operator.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across tasks: HASH (via `keyBy`)"/>
+<br/>
+<i><small>Figure 4: Message passing across tasks: HASH (keyBy)</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/hash.png)
+
+* A __REBALANCE__ distribution is either caused by an explicit call to `rebalance()` or by a change of parallelism (12 -> 1 in case of the job graph from Figure 2). Calling `rebalance()` causes data to be repartitioned randomly and can help to mitigate data skew in certain scenarios.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rebalance.png" width="800px" alt="Figure 4: Message passing across tasks: REBALANCE"/>
+<br/>
+<i><small>Figure 5: Message passing across tasks: REBALANCE</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/rebalance.png)
+
+At the top of the job graph in Figure 2 you can see an additional data source - **Rules Source**, which also consumes from Kafka. Rules are "mixed into" the main processing data flow through the `broadcast` channel. Unlike other methods of transmitting data between operators, such as `forward`, `hash` or `rebalance`, which make each message available for processing in only one of the parallel instances of the receiving operator, `broadcast` makes each message available at the input of all of the parallel instances of the operator which the _broadcast stream_ is connected to. This makes `broadcast` applicable to a wide range of tasks that need to affect the processing of all messages, regardless of their key or source partition.
```

Comment:
It is because there are a lot of images in the description - the section itself is 13 lines of text. I do not have a fresh idea neither how to restructure it nor how it explain it better. Sometimes people even have to flip the pages to see the relevant figure - seems unavoidable sometimes.

Reply:
Hmm...maybe you could re-use the jobgraph here and present the "Rules Source", "Transaction Source", and "Dynamic Partitioning" again as an example (just these three tasks and the arrows between them). Then, you could say that this is the same for the alert function and drop the code that creates this?